src/commands/login.rs

```diff
@@ -0,0 +1,400 @@
+        env = BINDLE_USERNAME,
+        requires = BINDLE_PASSWORD
+    )]
+    pub bindle_username: Option<String>,
+
+    /// Basic http auth password for the bindle server
+    #[clap(
+        name = BINDLE_PASSWORD,
+        long = "bindle-password",
+        env = BINDLE_PASSWORD,
+        requires = BINDLE_USERNAME
+    )]
+    pub bindle_password: Option<String>,
+
+    /// Ignore server certificate errors from bindle and hippo
+    #[clap(
+        name = INSECURE_OPT,
+        short = 'k',
+        long = "insecure",
+        takes_value = false,
+    )]
+    pub insecure: bool,
+
+    /// URL of hippo server
+    #[clap(
+        name = HIPPO_SERVER_URL_OPT,
+        long = "hippo-server",
+        env = HIPPO_URL_ENV,
+        requires = BINDLE_SERVER_URL_OPT,
+        requires = HIPPO_USERNAME,
+        requires = HIPPO_PASSWORD,
+    )]
+    pub hippo_server_url: Option<String>,
+
+    /// Hippo username
+    #[clap(
+        name = HIPPO_USERNAME,
+        long = "hippo-username",
+        env = HIPPO_USERNAME,
+        requires = BINDLE_SERVER_URL_OPT,
+        requires = HIPPO_SERVER_URL_OPT,
+        requires = HIPPO_PASSWORD,
+    )]
+    pub hippo_username: Option<String>,
+
+    /// Hippo password
+    #[clap(
+        name = HIPPO_PASSWORD,
+        long = "hippo-password",
+        env = HIPPO_PASSWORD,
+        requires = BINDLE_SERVER_URL_OPT,
+        requires = HIPPO_SERVER_URL_OPT,
+        requires = HIPPO_USERNAME,
+    )]
+    pub hippo_password: Option<String>,
+
+    /// Display login status
+    #[clap(name = "status", long = "status", takes_value = false)]
+    pub status: bool,
+
+    // fetch a device code
+    #[clap(
+        name = "get-device-code",
+        long = "get-device-code",
+        takes_value = false
+    )]
+    pub get_device_code: bool,
+
+    // check a device code
+    #[clap(
+        name = "check-device-code",
+        long = "check-device-code",
+        takes_value = false
+    )]
+    pub check_device_code: Option<String>,
+}
+
+impl LoginCommand {
+    pub async fn run(self) -> Result<()> {
+        let root = dirs::config_dir()
+            .context("Cannot find configuration directory")?
+            .join("spin");
+
+        ensure(&root)?;
+
+        let path = root.join("config.json");
+
+        if self.status {
+            let data = fs::read_to_string(path.clone())
+                .await
+                .context("Cannnot display login information")?;
+            println!("{}", data);
+            return Ok(());
+        }
+
+        let login_connection: LoginConnection;
+        let mut url = DEFAULT_CLOUD_URL.to_owned();
+        let mut auth_method = AuthMethod::Github;
+
+        if let Some(u) = self.hippo_server_url {
+            url = u;
+            // prompt the user for the authentication method
+            // TODO: implement a server "feature" check that tells us what authentication methods it supports
+            auth_method = prompt_for_auth_method();
```

Comment:
Sorry to be a broken record but there needs to be a way to bypass this for programmatic environments (e.g. so they can ask via a GUI and pass the setting down to `spin login`) please?

Also a couple of issues that you may already have in hand:

1. it's currently being shown even when `--get-device-code` or `--check-device-code` is present
2. currently if you provide a Hippo URL you're still required to provide username, password and Bindle server, even though you may be planning to answer "GitHub" at the prompt

Reply:
resolved:

- set to `AuthMethod::Github` if either `--get-device-code` or `--check-device-code` is set
- set to `AuthMethod::Github` if `--method github`
- set to `AuthMethod::UsernameAndPassword` if `--method username`